ext/timers/01_timers.js

```diff
@@ -61,534 +48,307 @@
-    // I'm not if that's possible.
-    opStartGlobalTimer(timeout);
-    await opWaitGlobalTimer();
-    pendingEvents--;
-    prepareReadyTimers();
-  }
+  /**
+   * A doubly linked list of timers.
+   * @type { { head: ScheduledTimer | null, tail: ScheduledTimer | null } } */
+  const scheduledTimers = { head: null, tail: null };
 
-  function prepareReadyTimers() {
-    const now = DateNow();
-    // Bail out if we're not expecting the global timer to fire.
-    if (globalTimeoutDue === null || pendingEvents > 0) {
-      return;
-    }
-    // After firing the timers that are due now, this will hold the first timer
-    // list that hasn't fired yet.
-    let nextDueNode;
-    while ((nextDueNode = dueTree.min()) !== null && nextDueNode.due <= now) {
-      dueTree.remove(nextDueNode);
-      // Fire all the timers in the list.
-      for (const timer of nextDueNode.timers) {
-        // With the list dropped, the timer is no longer scheduled.
-        timer.scheduled = false;
-        // Place the callback to pending timers to fire.
-        ArrayPrototypePush(pendingFireTimers, timer);
-      }
-    }
-    setOrClearGlobalTimeout(nextDueNode && nextDueNode.due, now);
-  }
+  /**
+   * @param {() => void} cb Will be run after the timeout, if it hasn't been
+   * cancelled.
+   * @param {number} millis
+   * @param {number} cancelRid
+   */
+  function runAfterTimeout(cb, millis, cancelRid) {
+    /** @type {ScheduledTimer} */
+    const timerObject = {
+      millis,
+      cb,
+      resolved: false,
+      prev: scheduledTimers.tail,
+      next: null,
+    };
 
-  function setOrClearGlobalTimeout(due, now) {
-    if (due == null) {
-      clearGlobalTimeout();
+    // Add timerObject to the end of the list.
+    if (scheduledTimers.tail === null) {
+      assert(scheduledTimers.head === null);
+      scheduledTimers.head = scheduledTimers.tail = timerObject;
     } else {
-      setGlobalTimeout(due, now);
+      scheduledTimers.tail.next = timerObject;
+      scheduledTimers.tail = timerObject;
     }
-  }
 
-  function schedule(timer, now) {
-    assert(!timer.scheduled);
-    assert(now <= timer.due);
-    // Find or create the list of timers that will fire at point-in-time `due`.
-    const maybeNewDueNode = { due: timer.due, timers: [] };
-    let dueNode = dueTree.find(maybeNewDueNode);
-    if (dueNode === null) {
-      dueTree.insert(maybeNewDueNode);
-      dueNode = maybeNewDueNode;
-    }
-    // Append the newly scheduled timer to the list and mark it as scheduled.
-    ArrayPrototypePush(dueNode.timers, timer);
-    timer.scheduled = true;
-    // If the new timer is scheduled to fire before any timer that existed before,
-    // update the global timeout to reflect this.
-    if (globalTimeoutDue === null || globalTimeoutDue > timer.due) {
-      setOrClearGlobalTimeout(timer.due, now);
-    }
-  }
+    // 1.
+    PromisePrototypeThen(
+      core.opAsync("op_sleep", millis, cancelRid),
```

Comment:
This means there's an op call per timer? To keep overhead down JS timers would ideally all hang off a single `tokio::time::sleep()`.

Reply:
The trade-off here is either:

a) Having at most one pending timer op at a time, but having to call into Rust to start another op every time one timer expires.
b) Only calling into Rust when starting or canceling timers, but having multiple pending timer ops.

I haven't measured the difference between these, and due to the very nature of timers measuring performance isn't easy. There's the additional confounding factor of #10974 if we compare with the current timers implementation. So I'm not sure how to benchmark this.

In any case, unrefing timers would be much easier with b) – see #12953.

Reply:
I agree that we should go with this approach, especially in light of ref/unref for timers.

Reply:
Node has regular JS timers hanging off a single C++ timer (N->1, common case) and unref'd JS timers backed by individual C++ timers (N->N, uncommon case.)

For Node, that's a worthwhile optimization because programs often have thousands of active timers, usually one or more per TCP connection.

For Deno, I'm not sure. TBH, I don't know how (or if) we deal with read/write timeouts or slowloris style attacks.

At any rate, it's an area worth investigating but it doesn't have to hold up this PR.